src/main/java/com/google/api/generator/engine/ast/CastExpr.java

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class CastExpr implements Expr {
+  @Override
+  public abstract TypeNode type();
+
+  public abstract Expr expr();
+
+  @Override
+  public void accept(AstNodeVisitor visitor) {
+    visitor.visit(this);
+  }
+
+  public static Builder builder() {
+    return new AutoValue_CastExpr.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setType(TypeNode type);
+
+    public abstract Builder setExpr(Expr expr);
+
+    abstract CastExpr autoBuild();
+
+    public CastExpr build() {
+      CastExpr castExpr = autoBuild();
+      Preconditions.checkState(
+          TypeNode.isReferenceType(castExpr.type()), "Casting can only be done to reference types");
+      Preconditions.checkState(
+          TypeNode.isReferenceType(castExpr.expr().type()),
+          "Casting can only be done on reference types");
+      return castExpr;
```

Comment:
Question on the two precondition check: how about the primitive type casting? like `int result = (int)(x / y);` Thanks!

Reply:
Good catch, fixed.